pkg/libvmi/vmi.go

```
@@ -290,7 +294,15 @@ func WithTPM(persistent bool) Option {
 }
 
 func baseVmi(name string) *v1.VirtualMachineInstance {
-	vmi := v1.NewVMIReferenceFromNameWithNS("", name)
+	vmi := &v1.VirtualMachineInstance{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: "",
+			SelfLink:  fmt.Sprintf("/apis/%s/namespaces/%s/virtualmachineinstances/%s", v1.GroupVersion.String(), "", name),
```

Can we drop this? I see this field is deprecated and should not be needed anymore.

Passing "" to fmt.Sprintf also looks suspicious.

@EdDev

---

The field `SelfLink` is deprecated, so it should be okay to remove it and also to remove the tests asserting it.

---

The tests are not just asserting the SelfLink field i think? So instead of removing the test as a whole, should i remove SelfLink from NewVMIReferenceFromNameWithNS at `vendor/kubevirt.io/api/core/v1/types.go`

---

No, please don't touch `NewVMIReferenceFromNameWithNS`. Just remove the use of the field `SelfLink` in your code.

---

Yep, will do that. I do have one question though, What should I do about the tests? I don't think those tests are just asserting selfLink and are rather used for checking, shutdown and deletion of domains, I think.

---

Thanks! I looked into it, I am still confused as to which parts do we exactly need to assert because as far as I can tell, the test compare the whole vmi type/struct/object. I see two ways to go about this:
1. Generate the vmi in the test code also using libvmi, which makes the tests pass
2.  If you could tell me which fields we are actually concerned about, then just  assert those fields.

Please tell me if there is any other way, that I have overlooked or not thought about and which one would be preferred?

---

Can you please point me to tests that fail if you do not set `SelfLink`?

In general, if you see e.g. `Expect(vmi.SelfLink).To(...)`, then simply remove that assertion. If full VMIs are compared, then remove `SelfLink` from the VMI that was not generated by libvmi.

---

Sorry for the delayed reply. 
The following tests fail:
https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/vm_test.go#L297
https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/vm_test.go#L307
https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/vm_test.go#L319
https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/vm_test.go#L337
https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/vm_test.go#L469
https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/vm_test.go#L488

All of these tests do not use libvmi to create VMI. There are no direct selfLink assertions. Just removing the selfLink before the expect works and passes the test successfully.
Should we not create the vmi in the test also using libvmi? That also solves the issue. May I know why this won't be a viable solution?

---

> Should we not create the vmi in the test also using libvmi? 

If that helsp to fix the issue then please change them too.

---

I have made the necessary changes and updated my PR. Please review.